### PR TITLE
Avoids allocation of map/slice of imported modules during instantiation

### PR DESCRIPTION
**Pull request by mathetake**
bench result will follow in a bit

### CODE REVIEW

File: internal/wasm/module.go [hunk 1]
```diff
@@ -48,6 +48,9 @@ type Module struct {
 	ImportGlobalCount,
 	ImportMemoryCount,
 	ImportTableCount Index
+	// ImportModuleNames hold the module names imported by this module. This is used during instantiation to mitigate
+	// the lock contention.
+	ImportModuleNames map[string]struct{}
```
Comment (achille-roussel) [inline internal/wasm/module.go:53]: It looks like this is only iterated on, maybe make it a `[]string`, could even be sorted for more determinism.

Comment (mathetake) [reply internal/wasm/module.go:53]: I thought about it first, but in anyway, we need to duplicate at binary decoding phrase to create the slice. So I decided to reuse the map here. But not a strong opinion. In fact, somehow current change seems not to reduce the allocation at all according to the perf result 😞 maybe something wrong here

Comment (mathetake) [reply internal/wasm/module.go:53]: using the slice resulted in the same.. 🤷

**Comment by mathetake** [comment]
ok looks like the `n := range ...` in requireModuleNames escape the names.

```go
	for n := range moduleNames {
		node, ok := s.nameToNode[n]
		if !ok {
			return nil, fmt.Errorf("module[%s] not instantiated", n)
		}
		ret[n] = node.module
	}
```

**Comment by mathetake** [comment]
ok changed the direction - removed the requireNames map and slice completely, and just get the imported module instances inside resolveImports. Let's see the bench...

**Comment by mathetake** [comment]
ok I see the increased lock contention... closing..

```
benchstat old.txt new.txt
goos: linux
goarch: amd64
pkg: github.com/tetratelabs/wazero/internal/integration_test/bench
cpu: AMD Ryzen 9 3950X 16-Core Processor
                                       │   old.txt   │               new.txt               │
                                       │   sec/op    │   sec/op     vs base                │
Initialization/interpreter-32            33.25µ ± 1%   32.26µ ± 0%   -2.98% (p=0.000 n=30)
Initialization/interpreter-multiple-32   14.16µ ± 0%   15.65µ ± 2%  +10.56% (p=0.000 n=30)
Initialization/compiler-32               38.60µ ± 1%   38.51µ ± 1%        ~ (p=0.187 n=30)
Initialization/compiler-multiple-32      13.65µ ± 1%   13.88µ ± 1%   +1.72% (p=0.000 n=30)
Compilation/with_extern_cache-32         292.4µ ± 1%   289.2µ ± 1%   -1.09% (p=0.020 n=30)
Compilation/without_extern_cache-32      10.61m ± 1%   10.82m ± 1%   +2.04% (p=0.000 n=30)
geomean                                  95.72µ        97.22µ        +1.58%

                                       │   old.txt    │               new.txt               │
                                       │     B/op     │     B/op      vs base               │
Initialization/interpreter-32            132.4Ki ± 0%   132.2Ki ± 0%  -0.19% (p=0.000 n=30)
Initialization/interpreter-multiple-32   132.9Ki ± 0%   132.6Ki ± 0%  -0.19% (p=0.000 n=30)
Initialization/compiler-32               133.1Ki ± 0%   132.9Ki ± 0%  -0.19% (p=0.000 n=30)
Initialization/compiler-multiple-32      137.4Ki ± 0%   137.1Ki ± 0%  -0.18% (p=0.000 n=30)
Compilation/with_extern_cache-32         53.76Ki ± 0%   53.76Ki ± 0%       ~ (p=0.589 n=30)
Compilation/without_extern_cache-32      1.907Mi ± 0%   1.907Mi ± 0%       ~ (p=0.139 n=30)
geomean                                  179.8Ki        179.6Ki       -0.13%

                                       │   old.txt   │              new.txt               │
                                       │  allocs/op  │  allocs/op   vs base               │
Initialization/interpreter-32             29.00 ± 0%    27.00 ± 0%  -6.90% (p=0.000 n=30)
Initialization/interpreter-multiple-32    42.00 ± 0%    40.00 ± 0%  -4.76% (p=0.000 n=30)
Initialization/compiler-32                29.00 ± 0%    27.00 ± 0%  -6.90% (p=0.000 n=30)
Initialization/compiler-multiple-32       32.00 ± 0%    30.00 ± 0%  -6.25% (p=0.000 n=30)
Compilation/with_extern_cache-32          986.0 ± 0%    986.0 ± 0%       ~ (p=0.334 n=30)
Compilation/without_extern_cache-32      32.62k ± 0%   32.62k ± 0%       ~ (p=0.186 n=30)
geomean                                   182.0         174.4       -4.18%

```

**Pull request closed without merging**